### PR TITLE
feat: handle federation errors when adding users to an MLS conversation WPB-2228

### PR DESCRIPTION
**Pull request by typfel**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-2228" title="WPB-2228" target="_blank">TaskWPB-2228</a>  [iOS] Implement MLS client protocol for adding users
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

When adding federated users (or creating a new group) to a conversation, the backends of those users might not be reachable or might not be properly configured to allow the users to be added. If this happens a client should remove exclude these users and insert a system informing the user about the users who couldn't be added.

This logic was already implemented for proteus conversations, this PR enables this logic to also be applied to MLS conversations by parsing and bubbling up federation errors during the MLS add participant action.

#### Test Coverage

- [x] I have added automated test to this contribution

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-request-strategy/Sources/Request Strategies/MLS/Actions/SendCommitBundleActionHandler.swift [hunk 1]
```diff
@@ -146,4 +174,13 @@ class SendCommitBundleActionHandler: ActionHandler<SendCommitBundleAction> {
...
         }
     }
+
+    // MARK: - Error response
+
+    struct FederationErrorResponse: Codable {
+
+        var unreachable_backends: [String]?
+        var non_federating_backends: [String]?
+
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Request Strategies/MLS/Actions/SendCommitBundleActionHandler.swift:184]:
```suggestion
        var unreachableBackends: [String]?
        var nonFederatingBackends: [String]?

        enum CodingKeys: String, CodingKey {
            case unreachableBackends = "unreachable_backends"
            case nonFederatingBackends = "non_federating_backends"
        }
```

**Review by netbe — CHANGES_REQUESTED** [review]
small suggestion + should it target develop now? I thought we merged feat/proteus-to-mls-migration already?

**Comment by typfel** [comment]
> small suggestion + should it target develop now? I thought we merged feat/proteus-to-mls-migration already?

yes it should, I'll update

**Comment by agisilaos** [comment]
> should it target develop now? I thought we merged feat/proteus-to-mls-migration already?

@typfel @netbe Indeed we merged it to develop but for some reason the branch wasn't deleted.

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/MLSConversationParticipantsService.swift [hunk 2]
```diff
@@ -36,6 +36,7 @@ protocol MLSConversationParticipantsServiceInterface {
 
 enum MLSConversationParticipantsError: Error, Equatable {
     case failedToClaimKeyPackages(users: Set<ZMUser>)
+    case failedAddUsers(users: Set<ZMUser>)
```
Comment (KaterinaWire) [inline wire-ios-request-strategy/Sources/Request Strategies/Conversation/MLSConversationParticipantsService.swift:39]: Where do we use this error?

**Review by David-Henner — APPROVED** [review]

**Comment by David-Henner** [comment]
Sorry! didn't mean to close the PR. 
@typfel could you update the target to `develop` ? I think then you'd be able to reopen the PR

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/MLSConversationParticipantsService.swift (hunk 2 above)
Comment (typfel) [reply wire-ios-request-strategy/Sources/Request Strategies/Conversation/MLSConversationParticipantsService.swift:39]: Not used, I'll delete

**Pull request closed without merging**